enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go

```diff
@@ -474,6 +478,58 @@ func (q *PriorityTaskScheduler) canFitTask(res *scpb.EnqueueTaskReservationReque
 	return true
 }
 
+// This function peeks at the front of the queue and checks to see if the task
+// still exists--if it doesn't, then we know the task was picked up *and
+// finished* by another worker, so we can safely prune it from the local queue,
+// allowing our queue size to go down even when we aren't picking up new tasks.
+//
+// During sudden bursts of large tasks, the task queues on individual executors
+// can get pretty long.  If we are autoscaling according to task queue length,
+// then doubling the number of executors might completely empty out the "real"
+// task queue, but if the original executor pool is fully stuck on large tasks,
+// the average task queue length will only be cut in half!  If that new length
+// is still over our autoscaling threshold, we'll scale up more even though we
+// don't need to.  Trimming makes this case less likely.
+//
+// This function returns true if a task was successfully dequeued.
+func (q *PriorityTaskScheduler) trimQueue() bool {
+	q.mu.Lock()
+	if q.shuttingDown || q.q.Len() == 0 {
+		return false
+	}
+	nextTask := q.q.Peek()
+	if nextTask == nil {
+		return false
+	}
+	// Hand the lock back in case this RPC takes a while.
+	q.mu.Unlock()
```

Comment:
missing unlock in the `return false` cases above - might be worth splitting into a separate func like `getQueuedTaskToPrune` so that we can defer the unlock